stdlib/universe/holt_winters.go

```diff
@@ -0,0 +1,463 @@
+	return t, d, nil
+}
+
+type holtWintersTransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+	alloc *fluxmemory.Allocator
+
+	withFit    bool
+	column     string
+	timeColumn string
+	n          int64
+	s          int64
+	interval   flux.Duration
+}
+
+func NewHoltWintersTransformation(d execute.Dataset, cache execute.TableBuilderCache, alloc *fluxmemory.Allocator, spec *HoltWintersProcedureSpec) *holtWintersTransformation {
+	return &holtWintersTransformation{
+		d:          d,
+		cache:      cache,
+		alloc:      alloc,
+		withFit:    spec.WithFit,
+		column:     spec.Column,
+		timeColumn: spec.TimeColumn,
+		n:          spec.N,
+		s:          spec.S,
+		interval:   spec.Interval,
+	}
+}
+
+func (hwt *holtWintersTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	// Sanity checks.
+	builder, created := hwt.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("holtWinters found duplicate table with key: %v", tbl.Key())
+	}
+	cols := tbl.Cols()
+	timeIdx := execute.ColIdx(hwt.timeColumn, cols)
+	if timeIdx < 0 {
+		return fmt.Errorf("cannot find time column %s", hwt.timeColumn)
+	}
+	colIdx := execute.ColIdx(hwt.column, cols)
+	if colIdx < 0 {
+		return fmt.Errorf("cannot find column %s", hwt.column)
+	}
+	typ := cols[colIdx].Type
+	if typ != flux.TInt &&
+		typ != flux.TUInt &&
+		typ != flux.TFloat {
+		return fmt.Errorf("holtWinters can work only on numerical types, got %s", typ.String())
+	}
+
+	// Building schema.
+	if err := execute.AddTableKeyCols(tbl.Key(), builder); err != nil {
+		return err
+	}
+	newTimeIdx, err := builder.AddCol(flux.ColMeta{
+		Label: execute.DefaultTimeColLabel,
+		Type:  flux.TTime,
+	})
+	if err != nil {
+		return err
+	}
+	newValueIdx, err := builder.AddCol(flux.ColMeta{
+		Label: execute.DefaultValueColLabel,
+		Type:  flux.TFloat,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Cleaning data for HoltWinters input.
+	vs, start, stop, err := hwt.getCleanData(tbl, colIdx, timeIdx)
+	if err != nil {
+		return err
+	}
+
+	// Holt Winters.
+	hw := holt_winters.New(int(hwt.n), int(hwt.s), hwt.withFit)
+	bf := func() *array.Float64Builder {
+		return array.NewFloat64Builder(fluxarrow.NewAllocator(hwt.alloc))
+	}
+	newVs := hw.Do(vs, bf)
+	// don't need vs anymore
+	vs.Release()
+
+	// Crafting timestamps.
+	// Timestamps are deduced by summing the interval to the first/last valid timestamp.
+	tsb := array.NewInt64Builder(fluxarrow.NewAllocator(hwt.alloc))
+	s := stop + int64(hwt.interval)
+	if hwt.withFit {
+		s = start
+	}
+	for i := 0; i < newVs.Len(); i++ {
+		tsb.Append(s)
+		s += int64(hwt.interval)
+	}
+	newTs := tsb.NewInt64Array()
+	defer func() {
+		newVs.Release()
+		newTs.Release()
+	}()
+
+	// Appending columns.
+	if err := builder.AppendTimes(newTimeIdx, newTs); err != nil {
+		return err
+	}
+	if err := builder.AppendFloats(newValueIdx, newVs); err != nil {
+		return err
+	}
+	for j, c := range tbl.Key().Cols() {
+		v := tbl.Key().Value(j)
```

Comment:
did you consider using  ` execute.AppendKeyValues(tbl.Key(), builder)`  for `newVS.Len()` iterations?   it's inside-out of how you append, perhaps yours is a bit faster?

Reply:
I was definitely looking for that function, thank you!